libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -899,3 +951,430 @@ float AP_GPS::get_lag(uint8_t instance) const
         return drivers[instance]->get_lag();
     }
 }
+
+/*
+ calculate the weightings used to blend GPs location and velocity data
+*/
+void AP_GPS::calc_blend_weights(void)
+{
+    // zero the blend weights
+    memset(&_blend_weights, 0, sizeof(_blend_weights));
+
+    // determine how many GPS receivers are returning a fix and the instance of the first receiver that is
+    uint8_t first_instance = 0;
+    for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
+        if (state[i].status > NO_FIX) {
+            num_instances = i+1;
+            if (first_instance == 0) {
+                first_instance = i;
+            }
+        }
+    }
+
+    // If not enough receivers to do blending or blending has  been disabled, set the primary instance and weighting to the first receiver detected
+    if (num_instances < 2 || _auto_switch != 2) {
+        primary_instance = first_instance;
+        _output_is_blended = false;
+        _blend_weights[primary_instance] = 1.0f;
+        return;
+    }
+
+    // Use the oldest non-zero time, but if time difference is excessive, use newest to prevent a disconnected receiver from blocking updates
+    uint32_t max_ms = 0; // newest non-zero system time of arrival of a GPS message
+    uint32_t min_ms = -1; // oldest non-zero system time of arrival of a GPS message
+    uint8_t newest_instance = 0; // index of receiver with newest time stamp
+    int16_t max_rate_ms = 0; // largest update interval of a GPS receiver
+    for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
+        // Find largest and smallest times
+        if (state[i].last_gps_time_ms > max_ms) {
+            max_ms = state[i].last_gps_time_ms;
+            newest_instance = i;
+        }
+        if ((state[i].last_gps_time_ms < min_ms) && (state[i].last_gps_time_ms > 0)) {
+            min_ms = state[i].last_gps_time_ms;
+        }
+        if (_rate_ms[i] > max_rate_ms) {
+            max_rate_ms = _rate_ms[i];
+        }
+    }
+    if ((int32_t)(max_ms - min_ms) < (int32_t)(2 * max_rate_ms)) {
+        // data is not too delayed so use the oldest time_stamp to give a chance for data from that receiver to be updated
+        state[GPS_MAX_RECEIVERS].last_gps_time_ms = min_ms;
+    } else {
+        // receiver data has timed out so hard switch to the newest data
+        state[GPS_MAX_RECEIVERS].last_gps_time_ms = max_ms;
+        // hard switch to that receiver
+        primary_instance = newest_instance;
```

Comment:
@rmackay9 I understand how it is choosing, I'm saying that doesn't look the best method to me. A 2D position isn't used by anyone (as far as I'm aware), but an old 3D position can still be used.
Again, this isn't so important now because we only have two GPS instances but when that number is increased, it can happen that it chooses a GPS with 2D fix when it has another with 3D fix that just received its last message one millisecond earlier (and so wasn't the chosen one).